core/consensus/slots/src/aux_schema.rs

```diff
@@ -52,67 +55,65 @@ pub struct EquivocationProof<H> {
 	snd_header: H,
 }
 
-impl<H> EquivocationProof<H> {
-	/// Get the slot number where the equivocation happened.
-	pub fn slot(&self) -> u64 {
-		self.slot
-	}
-
-	/// Get the first header involved in the equivocation.
-	pub fn fst_header(&self) -> &H {
-		&self.fst_header
-	}
-
-	/// Get the second header involved in the equivocation.
-	pub fn snd_header(&self) -> &H {
-		&self.snd_header
-	}
-}
-
 /// Checks if the header is an equivocation and returns the proof in that case.
 ///
 /// Note: it detects equivocations only when slot_now - slot <= MAX_SLOT_CAPACITY.
-pub fn check_equivocation<C, H, P>(
+pub fn check_equivocation<C, H, E, V>(
 	backend: &C,
 	slot_now: u64,
 	slot: u64,
 	header: &H,
-	signer: &P,
-) -> ClientResult<Option<EquivocationProof<H>>>
+	signature: V,
+	signer: &V::Signer,
+) -> ClientResult<Option<E>>
 	where
 		H: Header,
 		C: AuxStore,
-		P: Clone + Encode + Decode + PartialEq,
+		V: Verify + Codec + Clone,
+		<V as Verify>::Signer: Clone + Codec + PartialEq,
+		E: AuthorshipEquivocationProof<
+			Header=H,
+			Signature=V,
+			Identity=<V as Verify>::Signer,
+		>,
 {
 	// We don't check equivocations for old headers out of our capacity.
 	if slot_now - slot > MAX_SLOT_CAPACITY {
 		return Ok(None)
 	}
 
 	// Key for this slot.
-	let mut curr_slot_key = SLOT_HEADER_MAP_KEY.to_vec();
-	slot.using_encoded(|s| curr_slot_key.extend(s));
+	let mut current_slot_key = SLOT_HEADER_MAP_KEY.to_vec();
+	slot.using_encoded(|s| current_slot_key.extend(s));
 
 	// Get headers of this slot.
-	let mut headers_with_sig = load_decode::<_, Vec<(H, P)>>(backend, &curr_slot_key[..])?
-		.unwrap_or_else(Vec::new);
+	let mut headers_with_signature = load_decode::<_, Vec<(H, V, V::Signer)>>(
+		backend.deref(),
+		&current_slot_key[..],
+	)?
+	.unwrap_or_else(Vec::new);
 
 	// Get first slot saved.
 	let slot_header_start = SLOT_HEADER_START.to_vec();
 	let first_saved_slot = load_decode::<_, u64>(backend, &slot_header_start[..])?
 		.unwrap_or(slot);
 
-	for (prev_header, prev_signer) in headers_with_sig.iter() {
+	for (prev_header, prev_signature, prev_signer) in headers_with_signature.iter() {
 		// A proof of equivocation consists of two headers:
 		// 1) signed by the same voter,
 		if prev_signer == signer {
 			// 2) with different hash
 			if header.hash() != prev_header.hash() {
-				return Ok(Some(EquivocationProof {
-					slot, // 3) and mentioning the same slot.
-					fst_header: prev_header.clone(),
-					snd_header: header.clone(),
-				}));
+				return Ok(Some(AuthorshipEquivocationProof::new(
```

Comment:
Why not just use the existing struct and just pass the remaining data as arguments to the construct report call?

Reply:
Why not doing it in this way? Anyway the `proof` and the `SessionIndex` should be in the `Equivocation`, right?.